### PR TITLE
ARROW-1149: [Plasma] Create Cython client library for Plasma

### PR DESCRIPTION
**Pull request by pcmoritz**
This PR introduces a Cython API to Plasma, a FindPlasma.cmake to make it easier to integrate Plasma with CMake projects and sets up packaging with pyarrow.

### CODE REVIEW

**Comment by wesm** [comment]
@pcmoritz I'm sorry about the delay, I will have a closer look most likely tomorrow

**Comment by pcmoritz** [comment]
Thanks, looking forward to it! The central question is how much mixing between the plasma client and pyarrow we want and also how the directory structure should look like. I kept them pretty separate for now.

File: cpp/src/plasma/plasma.h
Comment (wesm) [inline cpp/src/plasma/plasma.h]: This is very odd, will have to have a closer look at what is going on; I've never had any issues like this

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: Mixed indentation -- use 4 spaces everywhere?

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: Raise a ValueError here instead

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: Add `with nogil:` here to release the GIL

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: `with nogil:` here (may require some unboxing of the Python objects outside the with nogil block)

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: `with nogil:`

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: Can you use numpydoc style here and in the other docstrings? https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: `with nogil`

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: with nogil

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: with nogil

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: with nogil

File: cpp/src/plasma/plasma/plasma.pyx
Comment (wesm) [inline cpp/src/plasma/plasma/plasma.pyx]: with nogil, and the other client calls below

File: cpp/src/plasma/test/test.py
Comment (wesm) [inline cpp/src/plasma/test/test.py]: Indent with 4 spaces in this file

**Review by wesm — COMMENTED** [review]
Thanks for doing this! As far as the packaging, it's a little more work to add a `FindPlasma.cmake` file to the pyarrow build system, but for packaging simplicity (for pip and conda) it would be easiest to maintain this as `pyarrow/plasma.pyx` and have a single build system and package artifact

I left minor comments about proper usage of `nogil` (you have to have the `with nogil:` block for it to release the GIL; the `nogil` annotation in the `cdef extern` blocks tells Cython that the functions are safe to use in a nogil block, but the actual release must be explicit)

The other thing is the docstrings so that things will render properly when we add these APIs to the Arrow Python documentation (http://arrow.apache.org/docs/python/api.html -- it's in need of some love as you can see)

**Comment by pcmoritz** [comment]
Thanks a lot for the review. The packaging simplicity is a good argument, let's make it part of pyarrow. I'm traveling right now but hope to get to it this weekend.

File: ci/travis_script_plasma.sh
Comment (robertnishihara) [inline ci/travis_script_plasma.sh]: probably comment this in?

File: python/pyarrow/tests/test_parquet.py
Comment (robertnishihara) [inline python/pyarrow/tests/test_parquet.py]: probably need an extra newline before `@parquet`

**Comment by xhochy** [comment]
@pcmoritz please ping me when I should review or you run into problems. I'm ignoring this pull request for now as there's a lot of activity in here with commits.

**Comment by pcmoritz** [comment]
@xhochy I'm wrapping it up right now, the last thing I'm really blocked on is the error in https://travis-ci.org/pcmoritz/arrow/jobs/253597521 in the manylinux build, see the error message
"ImportError: libplasma.so.0: cannot open shared object file: No such file or directory"

If you have an idea please let me know! I did add -DARROW_PLASMA=ON to `python/manylinux1/build_arrow.sh` and FindPlasma.cmake seems to find libplasma.so at least.

Other than that there are some doc fixes I need to do, I'll let you know when it is ready.

File: python/pyarrow/plasma.pyx
Comment (robertnishihara) [inline python/pyarrow/plasma.pyx]: In Ray, this used to look more like

```python
if self.plasma_client.alive:
    libplasma.release(self.plasma_client.conn, self.plasma_id)
```

I think the motivation for this was that Ray (on a single machine) by default kills the plasma store manually at the end of a script. After this, a `PlasmaBuffer` can still go out of scope and try to send a `release` message to the store, and the `check_status` will raise an exception leaving some ugly error or warning messages.

Or is this avoided in some other way?

File: cpp/src/plasma/CMakeLists.txt
Comment (xhochy) [inline cpp/src/plasma/CMakeLists.txt]: We could also move these `cmake_modules` to the top-level so that they are available to any `arrow-cpp` component.

File: python/manylinux1/build_arrow.sh
Comment (xhochy) [inline python/manylinux1/build_arrow.sh]: Can you add a check in line 67 below that plasma can be imported?

File: cpp/src/plasma/test/test.py
Comment (xhochy) [inline cpp/src/plasma/test/test.py]: Please specify valgrind and plasma options as part of the pytest call, see the `--parquet` option in `pyarrow/tests/conftest.py`

**Review by xhochy — CHANGES_REQUESTED** [review]

File: python/pyarrow/plasma.pyx
Comment (pcmoritz) [inline python/pyarrow/plasma.pyx]: This is now done in the C++ client; the file descriptor is set to -1 on disconnect, our old shutdown is disconnect and release checks for the file descriptor being -1.

Comment (robertnishihara) [inline python/pyarrow/plasma.pyx]: I see, that looks good.

File: cpp/src/plasma/CMakeLists.txt
Comment (pcmoritz) [inline cpp/src/plasma/CMakeLists.txt]: Yeah, I think that's where they belong. Shall we do this as a separate PR to make it a little more visible? It might break other people's scripts.

File: python/manylinux1/build_arrow.sh
Comment (pcmoritz) [inline python/manylinux1/build_arrow.sh]: Yes, thanks for the pointer!

File: .travis.yml
Comment (pcmoritz) [inline .travis.yml]: added tests for mac os now, if it becomes too burdensome, we can switch it off

**Comment by wesm** [comment]
Sorry about the rebase conflict in conftest.py, should not be hard to resolve. It looks like the `plasma` Cython module needs to be excluded from the Python build on MSVC

**Comment by pcmoritz** [comment]
The build is now green and it's ready to merge. One possible reservation is that travis times are quite long now, any thoughts on that?

**Comment by wesm** [comment]
I will give this a last look and merge later today assuming no issues now that the 0.5.0 release vote has passed. I will also look at the build times to see if there's anything that needs to be done

**Comment by pcmoritz** [comment]
Great, thanks! Next steps from our side is a little bit more high level documentation (I'll create a PR once this is merged) and then the blog post.

**Comment by wesm** [comment]
Sounds good. We can probably publish the blog post right after the IP clearance process is concluded

**Comment by wesm** [comment]
There's a significant amount of redundant work happening in the build. We should probably only be testing the built-in third-party toolchain (the ExternalProjects) in one entry in the build matrix and using toolchain libraries in all the others. This would probably be simplest to tackle in a follow up patch. I can take a look

File: python/doc/source/api.rst
Comment (wesm) [inline python/doc/source/api.rst]: Typo. Can address when working on documentation generally

File: python/pyarrow/includes/libarrow.pxd
Comment (wesm) [inline python/pyarrow/includes/libarrow.pxd]: Minor typo

File: python/pyarrow/io.pxi
Comment (wesm) [inline python/pyarrow/io.pxi]: Since buffers have an `is_mutable()` function, is this needed (versus adding a `mutable` property to Buffer)?

File: python/pyarrow/plasma.pyx
Comment (wesm) [inline python/pyarrow/plasma.pyx]: May want to move this to `io.pxi`

File: python/pyarrow/plasma.pyx
Comment (wesm) [inline python/pyarrow/plasma.pyx]: Let's see if this can be collapsed per the discussion above with `MutableBuffer`

File: python/pyarrow/tests/test_plasma.py
Comment (wesm) [inline python/pyarrow/tests/test_plasma.py]: What do you think about adding a top level function `pyarrow.plasma.connect` that returns a fully-baked client? This API is a slight rough edge

**Review by wesm — COMMENTED** [review]
Minor comments -- sorry to delay getting this in, but I think we should remove the `MutableBuffer` classes and instead have a `mutable` property that forwards `buffer->is_mutable()`. I can look into speeding up the builds in a separate patch

File: python/pyarrow/tests/test_plasma.py
Comment (pcmoritz) [inline python/pyarrow/tests/test_plasma.py]: Yeah, you are right, that seems better!

File: python/setup.py
Comment (pcmoritz) [inline python/setup.py]: The function initialize_options seems to have been called twice and plasma was added twice to CYTHON_MODULE_NAMES, which broke the "python setup.py develop". If there is a better fix let me know!

File: python/setup.py
Comment (pcmoritz) [inline python/setup.py]: I added this for good measure

File: python/pyarrow/plasma.pyx
Comment (wesm) [inline python/pyarrow/plasma.pyx]: As one rough edge in Cython, if you call the ctor with the wrong parameters you will generally segfault, which is why factory methods are usually safer. We can tackle this in a follow up patch

**Review by wesm — APPROVED** [review]
+1, thanks @pcmoritz!!

**Comment by pcmoritz** [comment]
Thanks for merging! If you could look into the ExternalProjects build times that'd be great (I'm not sure what is involved there). I can see if we can make the actual plasma tests a little faster (and can also change the PlasmaClient -> connect API).

**Pull request closed without merging**
